### PR TITLE
[TECH-223] Upgrade MediatR to 12.x

### PR DESCRIPTION
**Pull request by asimmon**
* Followed [MediatR 11.x to 12.x migration guide](https://github.com/jbogard/MediatR/wiki/Migration-Guide-11.x-to-12.0)
* Fixed a few ~bugs~ incomplete implementations
* Added more test coverage

### CODE REVIEW

File: src/GSoft.Extensions.MediatR.ApplicationInsights/RequestApplicationInsightsBehavior.cs [hunk 1]
```diff
@@ -1,11 +1,12 @@
...
+﻿using System.Diagnostics;
+using MediatR;
 using Microsoft.ApplicationInsights;
 using Microsoft.ApplicationInsights.DataContracts;
 
 namespace GSoft.Extensions.MediatR;
 
 internal sealed class RequestApplicationInsightsBehavior<TRequest, TResponse> : IPipelineBehavior<TRequest, TResponse>
-    where TRequest : IRequest<TResponse>
+    where TRequest : notnull
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR.ApplicationInsights/RequestApplicationInsightsBehavior.cs:9]: This change was required due to a breaking change in MediatR 12.
See: https://github.com/jbogard/MediatR/wiki/Migration-Guide-11.x-to-12.0

File: src/GSoft.Extensions.MediatR.Tests/GSoft.Extensions.MediatR.Tests.csproj [hunk 2]
```diff
@@ -15,6 +15,7 @@
   <ItemGroup>
     <PackageReference Include="GSoft.Extensions.Xunit" Version="1.0.1" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.4.1" />
+    <PackageReference Include="Microsoft.TestPlatform.ObjectModel" Version="17.4.1" Condition=" '$(OS)' != 'Windows_NT' " />
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR.Tests/GSoft.Extensions.MediatR.Tests.csproj:18]: Required to run tests on `net462` on Linux. `net462` validates the library on `netstandard2.0`.

File: src/GSoft.Extensions.MediatR.Tests/InMemoryLoggerTracker.cs [hunk 3]
```diff
@@ -53,11 +56,15 @@ public void AssertRequestFailed(string startsWith)
...
-            Assert.Single(this._logs, x => x.StartsWith(startsWith + " started"));
-            Assert.Single(this._logs, x => x.StartsWith(startsWith + " failed after"));
+            Assert.StartsWith(startsWith + " started", this._logs[0].Text);
+            Assert.StartsWith(startsWith + " failed after", this._logs[1].Text);
+            Assert.NotNull(this._logs[0].ActivityId);
+            Assert.Equal(this._logs[0].ActivityId, this._logs[1].ActivityId);
         }
     }
 
+    private sealed record LogRecord(string Text, string? ActivityId);
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR.Tests/InMemoryLoggerTracker.cs:66]: There was a bug, we now make sure that the OpenTelemetry span (activity) is the same for both start and end logs.

File: src/GSoft.Extensions.MediatR.Tests/IsExternalInit.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+﻿#if NETFRAMEWORK
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR.Tests/IsExternalInit.cs:1]: Required to use `init` properties in this tests targeting .NET 4.6.2

File: src/GSoft.Extensions.MediatR.Tests/MediatorTests.cs [hunk 5]
```diff
@@ -23,121 +23,157 @@ public MediatorTests(MediatorFixture fixture, ITestOutputHelper testOutputHelper
     }
 
     [Fact]
-    public async Task Send_Request_Works_When_Handler_Is_Succesful()
+    public async Task Send_Query_Works_When_Handler_Is_Succesful()
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR.Tests/MediatorTests.cs:26]: Now using CQRS naming conventions in preparation for TECH-224

File: src/GSoft.Extensions.MediatR/GSoft.Extensions.MediatR.csproj [hunk 6]
```diff
@@ -10,17 +10,16 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="MediatR" Version="11.1.0" />
-    <PackageReference Include="MediatR.Extensions.Microsoft.DependencyInjection" Version="11.0.0" />
+    <PackageReference Include="MediatR" Version="12.0.1" />
+    <PackageReference Include="MediatR.Contracts" Version="2.0.1" />
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR/GSoft.Extensions.MediatR.csproj:14]: I discovered that this package is manually published by jbogard, and he already made a mistake two weeks ago.

Although it's implicitely referenced by `MediatR` package, the `MediatR` still referenced the bugged and unlisted previous version `MediatR.Contracts`, so we need to explicitely reference `MediatR.Contracts` to better control our dependencies.

File: src/GSoft.Extensions.MediatR/GSoft.Extensions.MediatR.csproj [hunk 7]
```diff
@@ -30,9 +29,7 @@
...
     <InternalsVisibleTo Include="GSoft.Extensions.MediatR.Tests,PublicKey=002400000480000094000000060200000024000052534131000400000100010025301ce547647ab5ac9264ade0f9cdc0252796a257095add4791b0232c1def21bb9e0c87d218713f918565b23394362dbcb058e210c853a24ec33e6925ebedf654a0d65efb3828c855ff21eaaa67aeb9b24b81b8baff582a03df6ab04424c7e53cacbfe84d2765ce840389f900c55824d037d2c5b6b330ac0188a06ef6869dba" />
-    <InternalsVisibleTo Include="DynamicProxyGenAssembly2, PublicKey=0024000004800000940000000602000000240000525341310004000001000100c547cac37abd99c8db225ef2f6c8a3602f3b3606cc9891605d02baa56104f4cfc0734aa39b93bf7852f7d9266654753cc297e7d2edfe0bac1cdcf9f717241550e0a7b191195b7667bb4f64bcb8e2121380fd1d9d46ad2d92d2d15605093924cceaf74c4861eff62abf69b9291ed0a340e113be11e6a7d3113e92484cf7045cc7" />
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR/GSoft.Extensions.MediatR.csproj:36]: Wasn't necessary, we don't use Moq or FakeItEasy

File: src/GSoft.Extensions.MediatR/MediatorBuilder.cs [hunk 8]
```diff
@@ -1,54 +1,87 @@
 using System.Reflection;
 using MediatR;
 using Microsoft.Extensions.DependencyInjection;
-using Microsoft.Extensions.DependencyInjection.Extensions;
 
 namespace GSoft.Extensions.MediatR;
 
 public sealed class MediatorBuilder
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR/MediatorBuilder.cs:7]: Lots of changes here due to MediatR 12.x breaking changes about how to add the library.
Hopefully these breaking change will be abstracted by us in this library, so consumers won't be impacted in the future.

File: src/GSoft.Extensions.MediatR/MediatorExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
+﻿using System.Diagnostics;
+
+// ReSharper disable once CheckNamespace
+namespace MediatR;
+
+// These extension methods follows Microsoft guidelines in terms of asynchronous programming
+// and also encourage the developer to provide the right cancellation token depending on the context.
+// https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/async-scenarios#important-info-and-advice
+[DebuggerStepThrough]
+public static class MediatorExtensions
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR/MediatorExtensions.cs:10]: These methods will be suggested over non "Async" suffixed methods in the future

File: src/GSoft.Extensions.MediatR/RequestLoggingBehavior.cs [hunk 10]
```diff
@@ -28,14 +29,38 @@ public async Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TRe
...
-            this._logger.RequestFailed(ex, requestName, watch.Elapsed.TotalSeconds);
+
+            if (originatingActivity == null)
+            {
+                this._logger.RequestFailed(ex, requestName, watch.Elapsed.TotalSeconds);
+            }
+            else
+            {
+                // Make sure the logs being sent are attached to the originating activity
+                originatingActivity.ExecuteAsCurrentActivity(new FailedLoggerState(this._logger, requestName, watch.Elapsed.TotalSeconds, ex), static x =>
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR/RequestLoggingBehavior.cs:59]: This was a bug, ending logs were not attached to the same activity of the starting logs. Fixed + changed the unit tests

File: src/GSoft.Extensions.MediatR.ApplicationInsights/DictionaryExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+﻿#if NETSTANDARD2_0
```
Comment (asimmon) [inline src/GSoft.Extensions.MediatR.ApplicationInsights/DictionaryExtensions.cs:1]: Simply doesn't exists on netstandard2.0, but exists on net6.0.

**Review by Gcaya — APPROVED** [review]

File: src/GSoft.Extensions.MediatR.ApplicationInsights/RequestApplicationInsightsBehavior.cs [hunk 12]
```diff
@@ -21,22 +22,38 @@ public Task<TResponse> Handle(TRequest request, RequestHandlerDelegate<TResponse
...
+            operation.Telemetry.Success = false;
+            operation.Telemetry.Properties.TryAdd(ApplicationInsightsConstants.Exception, ex.ToString());
 
-                return result;
+            throw;
+        }
+        finally
+        {
+            // The dependency telemetry is sent when the operation is disposed
+            if (originatingActivity == null)
```
Comment (Zav) [inline src/GSoft.Extensions.MediatR.ApplicationInsights/RequestApplicationInsightsBehavior.cs:49]: Can this be null? (this code path don't seems to be covered in the tests)

Comment (asimmon) [reply src/GSoft.Extensions.MediatR.ApplicationInsights/RequestApplicationInsightsBehavior.cs:49]: Yes `Activity.Current` can be return null if there is no current activity being observed.

If we want to test this path we need to duplicate the current test class and fixture, then remove InMemoryActivityTracker from the fixture so we don't observe Activities.

**Pull request merged**